feat(manager): add sveltos manager

oliverbaehler opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->
This adds the manager for [sveltos](https://projectsveltos.github.io/sveltos/) for GitOps managed Cluster Fleets. Essentially we have the same functionality as argoCD, so we were able to take the argocd manager and adjust the schema etc. I hope that's okay.

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

Not sure if there's more?

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=30087) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=30087) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=30087) it.</sub>

secustor commented:
Please create first a discussion with the 'new package manager' label and the questionnaire filled out, so we have all information to make decision. 

I'm not closing this, but moving it into draft till then as it looks well crafted on a first glance.

oliverbaehler commented:
@secustor thanks for your feedback, i hope i have correctly created the discussion: https://github.com/renovatebot/renovate/discussions/30090

However i cant add the mentioned label, do you need to add it?

secustor commented:
Package manager details can be found here: https://github.com/renovatebot/renovate/discussions/30090#discussioncomment-9990664

secustor commented on lib/modules/manager/sveltos/extract.spec.ts:
```diff
@@ -0,0 +1,227 @@
+
+    it('return result for double quoted argoproj.io apiVersion reference', () => {
+      const result = extractPackageFile(
+        `
```
```suggestion
        codeblock`
```
Use `codeblock` to allow indentation of the inline fixtures. 
https://github.com/renovatebot/renovate/blob/057049b134f7987e5029f435f3d2a797bdccd04f/lib/util/toml.spec.ts#L6

secustor commented on lib/modules/manager/sveltos/__fixtures__/validEventTrigger.yml:
```diff
@@ -0,0 +1,48 @@
+    chartVersion:     "v3.2.5"
+---
+apiVersion: lib.projectsveltos.io/v1beta1
+kind: EventTrigger
```
Could people want to create rules specific to `EventTrigger`, `ClusterProfile` or `Profile`?

oliverbaehler replied:
It's probably more common to have them all combined instead of making specific rules. Should i combine the fixtures?

secustor replied:
I'm referring here to `depType`s. See this section of the `terraform` manager as an example. https://docs.renovatebot.com/modules/manager/terraform/#disabling-parts-of-the-manager

secustor commented on lib/modules/manager/sveltos/index.ts:
```diff
@@ -0,0 +1,21 @@
+
+export const supportedDatasources = [
+  DockerDatasource.id,
+  GitTagsDatasource.id,
```
```suggestion
```
I don't see any usages of this currently.

secustor commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,89 @@
+}
+
+function processAppSpec(
+  definition: SveltosHelmSource,
```
```suggestion
  definition: ProfileDefinition,
```

secustor commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,89 @@
+    deps.push(processSource(spec.source));
+  }
+
+  for (const source of coerceArray(spec.sources)) {
```
Where is this sources coming from?
I think this whole function needs a rework.

oliverbaehler replied:
Copied from ArgoCD https://github.com/renovatebot/renovate/blob/d65effaa77b53ef93312474e81fdbd0efeff0498/lib/modules/manager/argocd/extract.ts#L81

Do you have any hints what needs to be changed (I am a ts noob) :D

secustor replied:
You are trying to access a field which does not exist on the object. 
`spec.source` and `spec.sources` references these fields: 
- https://github.com/renovatebot/renovate/blob/d65effaa77b53ef93312474e81fdbd0efeff0498/lib/modules/manager/argocd/__fixtures__/validApplication.yml#L66
- https://github.com/renovatebot/renovate/blob/d65effaa77b53ef93312474e81fdbd0efeff0498/lib/modules/manager/argocd/__fixtures__/validApplication.yml#L82

secustor reviewed (changes requested):
Also needs prettier run. Check the contributor docs https://docs.renovatebot.com/contributing-to-renovate/

oliverbaehler commented:
@secustor Implemented suggestions

secustor commented:
Needs fixes and do not use force push as we squash the commits anyway while merging. 
Further please use the rereview function instead of tagging.

oliverbaehler commented:
Sorry, forced pushed again. Won't do again!

oliverbaehler commented:
@secustor Yes this is ready for review, Sorry!

HonkingGoose commented on lib/modules/manager/sveltos/readme.md:
```diff
@@ -0,0 +1,39 @@
+The [Sveltos](https://projectsveltos.github.io/sveltos/) manager allows to manage the dependencies of Helm-Charts for Sveltos resources. Read more about Sveltos Helm-Charts in the [Sveltos documentation](https://projectsveltos.github.io/sveltos/addons/helm_charts/).
+
+### Set `fileMatch` pattern
```
```suggestion
### You must set a `fileMatch` pattern
```

Put the key info in the heading: the reader must do something.

viceice replied:
we've other managers with empty file match, we should align them to use same layout

HonkingGoose commented on lib/modules/manager/sveltos/readme.md:
```diff
@@ -0,0 +1,39 @@
+### Set `fileMatch` pattern
+
+The `sveltos` manager has no default `fileMatch` pattern, because there is no common filename or directory name convention for Sveltos YAML files. By setting your own `fileMatch` Renovate avoids having to check each `*.yaml` file in a repository for a Sveltos definition.
+
```
```suggestion
#### fileMatch pattern examples

```

HonkingGoose commented on lib/modules/manager/sveltos/readme.md:
````diff
@@ -0,0 +1,39 @@
+}
+```
+
+### Disabling parts of the manager
````
```suggestion
### Disabling parts of the sveltos manager
```

HonkingGoose commented on lib/modules/manager/sveltos/readme.md:
```diff
@@ -0,0 +1,39 @@
+You can use these `depTypes` for fine-grained control, for example to disable parts of the Sveltos manager.
+
+| Resource                                                                                |    `depType`     |
+| --------------------------------------------------------------------------------------- | :--------------: |
```
```suggestion
| :-------------------------------------------------------------------------------------- | :--------------- |
```

Align all text to the left.

secustor reviewed (changes requested):
Needs code coverage ( see the failing CI ) and docs changes requested by `@HonkingGoose`

oliverbaehler commented on lib/modules/manager/sveltos/schema.ts:
```diff
@@ -0,0 +1,43 @@
+import { z } from 'zod';
+import { LooseArray } from '../../../util/schema-utils';
+
+export const KubernetesResource = z.object({
```
No, i have extended the object. Is that okay or not?

viceice replied:
we should unify it

https://github.com/renovatebot/renovate/blob/302718f38e54c6c7590f488d749fb4da8362f36f/lib/modules/manager/flux/schema.ts#L3-L13

viceice commented on lib/modules/manager/sveltos/extract.spec.ts:
```diff
@@ -0,0 +1,227 @@
+const validProfile = Fixtures.get('validProfile.yml');
+const validClusterProfile = Fixtures.get('validClusterProfile.yml');
+const validEventTrigger = Fixtures.get('validEventTrigger.yml');
+const malformedProfiles = Fixtures.get('malformedProfiles.yml');
```
move to test if only used once (perf)

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,89 @@
+): PackageFileContent | null {
+  let definitions: Record<string, unknown>[] | ProfileDefinition[];
+  try {
+    definitions = parseYaml(content, null, {
```
```suggestion
    definitions = parseYaml(content, {
```

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,89 @@
+export function extractPackageFile(
+  content: string,
+  packageFile: string,
+  _config?: ExtractConfig,
```
we should support `config.registryAliases`.

https://github.com/renovatebot/renovate/blob/302718f38e54c6c7590f488d749fb4da8362f36f/lib/modules/manager/flux/extract.ts#L150-L177

secustor commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,89 @@
+  try {
+    definitions = parseYaml(content, {
+      customSchema: ProfileDefinition,
+    }) as any;
```
```suggestion
    });
```
Do not use any

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,89 @@
+  packageFile: string,
+  _config?: ExtractConfig,
+): PackageFileContent | null {
+  let definitions: Record<string, unknown>[] | ProfileDefinition[];
```
```suggestion
  let definitions: ProfileDefinition[];
```

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,96 @@
+    return null;
+  }
+
+  const deps = definitions.flatMap((definition) => {
+    // Use zod's safeParse method to check if the object matches the ProfileDefinition schema
+    const result = ProfileDefinition.safeParse(definition);
+    if (result.success) {
+      return processAppSpec(result.data, config);
+    }
+    return [];
```
extract util function

oliverbaehler replied:
refactored the function (i hope that was you were talking about)

secustor commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,102 @@
+): PackageFileContent | null {
+  let definitions: ProfileDefinition[];
+  try {
+    definitions = parseYaml(content, null, {
```
```suggestion
    definitions = parseYaml(content, {
```
This still need to be updated

secustor reviewed (commented):
You are missing coverage.

oliverbaehler commented:
Hi I am sorry I am unable to get the coverage up :( . Can I somehow create the coverage locally? What i currently do is run Jest:

```
pnpm jest lib/modules/manager/sveltos
...
Statements   : 2.39% ( 1023/42686 )
Branches     : 0.36% ( 49/13404 )
Functions    : 0.51% ( 27/5294 )
Lines        : 2.42% ( 1020/42083 )
```

But i don't know how to interpret these results..

rarkins commented:
https://github.com/renovatebot/renovate/blob/9b64288861fcebfcce7a4938307dec52c04ec54f/docs/development/local-development.md?plain=1#L157-L166

viceice commented on lib/modules/manager/sveltos/extract.spec.ts:
```diff
@@ -1,5 +1,7 @@
 import { codeBlock } from 'common-tags';
 import { Fixtures } from '../../../../test/fixtures';
+import { extractDefinition } from './extract';
+import { ProfileDefinition } from './schema';
```
```suggestion
import type { ProfileDefinition } from './schema';
```

viceice commented on lib/modules/manager/sveltos/extract.spec.ts:
```diff
@@ -0,0 +1,308 @@
+      `;
+
+      jest
+        .spyOn(require('./extract'), 'extractDefinition')
```
wrong, doesn't work this way

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,102 @@
+  definition: ProfileDefinition,
+  config?: ExtractConfig,
+): PackageDependency[] {
+  const result = ProfileDefinition.safeParse(definition);
```
why parsing again? makes no sense as it's already validated in caller. remove and inline this function.

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,102 @@
+    return null;
+  }
+
+  const deps = definitions.flatMap((definition) =>
+    extractDefinition(definition, config),
+  );
```
use for-of loop

oliverbaehler replied:
adjusted, recheck

rarkins commented on .github/workflows/undesirable-test-additions.yaml:
```diff
@@ -15,6 +15,7 @@ jobs:
           fetch-depth: 0 # Fetch all history for comparison
           sparse-checkout: true
           ref: ${{ github.event.pull_request.head.ref }}
+          repository: ${{ github.event.pull_request.head.repo.full_name }}
```
```suggestion
```

HonkingGoose commented on lib/modules/manager/sveltos/extract.spec.ts:
```diff
@@ -0,0 +1,304 @@
+      expect(result).toBeNull();
+    });
+
+    it('return null if yaml is invalid', () => {
```
```suggestion
    it('return null if YAML is invalid', () => {
```

HonkingGoose commented on lib/modules/manager/sveltos/readme.md:
```diff
@@ -0,0 +1,45 @@
+This is because there is are no common filename or directory name conventions for Sveltos YAML files.
+You must set your own `fileMatch` rules, so Renovate knows which `*.yaml` files are Sveltos definitions.
+
+#### fileMatch pattern examples
```
```suggestion
#### `fileMatch` pattern examples
```

Let's use the `code` styling for the config option name here too.

secustor commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,101 @@
+
+  const depType = definition.kind;
+
+  if (spec && is.nonEmptyArray(spec.helmCharts)) {
+    for (const source of coerceArray(spec.helmCharts)) {
+      const dep = processHelmCharts(source, config?.registryAliases);
+      if (dep) {
+        dep.depType = depType;
+        deps.push(dep);
+      }
+    }
+  }
```
```suggestion
    for (const source of coerceArray(definition.spec?.helmCharts)) {
      const dep = processHelmCharts(source, config?.registryAliases);
      if (dep) {
        dep.depType = depType;
        deps.push(dep);
      }
    }
```
Simplification with optional chaining.

secustor commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,101 @@
+  definition: ProfileDefinition,
+  config?: ExtractConfig,
+): PackageDependency[] {
+  const spec = definition.spec;
+
```
```suggestion
```

secustor commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,101 @@
+
+  if (
+    isOCIRegistry(source.repositoryURL) ||
+    !source.repositoryURL.includes('://')
```
When would this be the case?

oliverbaehler replied:
Never, leftover code

oliverbaehler commented:
HI @secustor, I implemented your changes. I also moved the fixtures from the `__fixtures__` directory to inline consts since @rarkins  added a github check, that. fixtures no longer should be used. Trying to run:

```
Error: Cannot find module '@opentelemetry/semantic-conventions/incubating'
Require stack:
- /usr/src/app/lib/instrumentation/index.ts
- /usr/src/app/lib/renovate.ts
```

Lint also throws errors:

```
docker run -it --rm -v "${PWD}:/usr/src/app" -e LOG_LEVEL=debug -w /usr/src/app renovatebot_local  pnpm eslint . --cache --cache-location .cache/eslint --report-unused-disable-directives

> renovate@0.0.0-semantic-release eslint /usr/src/app
> eslint . --cache --cache-location .cache/eslint --report-unused-disable-directives "." "--cache" "--cache-location" ".cache/eslint" "--report-unused-disable-directives"


/usr/src/app/lib/expose.cjs
  1:1  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/no-require-imports')

/usr/src/app/lib/instrumentation/index.ts
  25:3   error  ATTR_SERVICE_NAME not found in '@opentelemetry/semantic-conventions'               import/named
  26:3   error  ATTR_SERVICE_VERSION not found in '@opentelemetry/semantic-conventions'            import/named
  28:40  error  Unable to resolve path to module '@opentelemetry/semantic-conventions/incubating'  import/no-unresolved

/usr/src/app/lib/modules/manager/pipenv/artifacts.ts
  1:40  error  Unable to resolve path to module '@renovatebot/detect-tools'  import/no-unresolved

/usr/src/app/lib/modules/manager/pipenv/extract.ts
  1:40  error  Unable to resolve path to module '@renovatebot/detect-tools'  import/no-unresolved

/usr/src/app/lib/modules/platform/bitbucket-server/index.ts
  731:3  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/only-throw-error')

/usr/src/app/lib/util/git/index.ts
  106:3  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/only-throw-error')

/usr/src/app/lib/util/host-rules.ts
  113:17  error  Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`), as it is a safer operator  @typescript-eslint/prefer-nullish-coalescing

/usr/src/app/lib/util/mutex.ts
  1:57  error  Unable to resolve path to module 'async-mutex'  import/no-unresolved

/usr/src/app/lib/util/result.spec.ts
  1:1  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/prefer-promise-reject-errors')
  2:1  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/only-throw-error')

/usr/src/app/lib/util/result.ts
  327:7  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/only-throw-error')
  352:7  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/only-throw-error')
  368:5  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/only-throw-error')
  383:7  error  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/only-throw-error')

/usr/src/app/lib/util/yaml.ts
   8:61  error  Unable to resolve path to module 'yaml'                    import/no-unresolved
  30:27  error  'any' overrides all other types in this intersection type  @typescript-eslint/no-redundant-type-constituents
  31:3   error  'any' overrides all other types in this intersection type  @typescript-eslint/no-redundant-type-constituents
  32:3   error  'any' overrides all other types in this intersection type  @typescript-eslint/no-redundant-type-constituents
  33:3   error  'any' overrides all other types in this intersection type  @typescript-eslint/no-redundant-type-constituents
  34:3   error  'any' overrides all other types in this intersection type  @typescript-eslint/no-redundant-type-constituents
```

DO I need to update/change something?

secustor commented:
Looks like you forgot to install the updated dependencies, after merging in main.

oliverbaehler commented:
@secustor Oh thanks! Sorry really not trying to waste everyone's time.. 3:

HonkingGoose commented on lib/modules/manager/sveltos/readme.md:
```diff
@@ -0,0 +1,45 @@
+| :-------------------------------------------------------------------------------------- | :--------------- |
+| [Cluster Profiles](https://projectsveltos.github.io/sveltos/addons/clusterprofile/)     | `ClusterProfile` |
+| [Profiles](https://projectsveltos.github.io/sveltos/addons/profile/)                    | `Profile`        |
+| [EventTrigger](https://projectsveltos.github.io/sveltos/events/addon_event_deployment/) | `EventTrigger`   |
```
```suggestion
| [EventTrigger](https://projectsveltos.github.io/sveltos/events/addon_event_deployment/#eventtrigger) | `EventTrigger`   |
```
Should we link to the section about `EventTrigger`s? Right now we're linking to the _general_ page for Events.

secustor reviewed (commented):
needs prettier fix, else LGTM

secustor reviewed (approved):
Thanks for your contribution!

oliverbaehler commented:
@secustor Thanks for merging, just a quick question: Does it take some time to update the version on developer.mend.io?

It keeps telling me that the configuration with the sveltos key is invalid 🤔:

```
INFO: Repository has invalid config
{
  "error": {
    "validationSource": "renovate.json",
    "validationError": "The renovate configuration file contains some invalid settings",
    "validationMessage": "Invalid configuration option: sveltos",
    "message": "config-validation",
    "stack": "Error: config-validation\n    at mergeRenovateConfig (/usr/local/renovate/lib/workers/repository/init/merge.ts:198:19)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at getRepoConfig (/usr/local/renovate/lib/workers/repository/init/config.ts:14:12)\n    at initRepo (/usr/local/renovate/lib/workers/repository/init/index.ts:56:12)\n    at Object.renovateRepository (/usr/local/renovate/lib/workers/repository/index.ts:64:14)\n    at attributes.repository (/usr/local/renovate/lib/workers/global/index.ts:202:11)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:187:7)\n    at /usr/local/renovate/lib/renovate.ts:18:22"
  }
}
```

secustor commented:
The hosted app is not updated automatically. Not sure when Mend will update the hosted app again.

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,87 @@
+  return deps.length ? { deps } : null;
+}
+
+export function extractDefinition(
+  definition: ProfileDefinition,
+  config?: ExtractConfig,
+): PackageDependency[] {
+  return processAppSpec(definition, config);
+}
```
wonder why we need this simple wrapper?

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,87 @@
+  };
+
+  if (isOCIRegistry(source.repositoryURL)) {
+    const image = trimTrailingSlash(removeOCIPrefix(source.repositoryURL));
```
why trim trailing slash? doesn't make sense to me

viceice commented on lib/modules/manager/sveltos/extract.ts:
```diff
@@ -0,0 +1,87 @@
+  return dep;
+}
+
+function processAppSpec(
```
this function should be inlined

viceice commented on lib/modules/manager/sveltos/util.ts:
```diff
@@ -0,0 +1,13 @@
+  const repoNameWithSlash = regEx(`^${repositoryName}/`, undefined, false);
+  let modifiedChartName = chartName.replace(repoNameWithSlash, '');
+
+  modifiedChartName = modifiedChartName.replace(/\/+$/, '');
```
same here, makes no sense and has no tests.

viceice commented on lib/modules/manager/sveltos/util.ts:
```diff
@@ -0,0 +1,13 @@
+  chartName: string,
+): string {
+  const repoNameWithSlash = regEx(`^${repositoryName}/`, undefined, false);
+  let modifiedChartName = chartName.replace(repoNameWithSlash, '');
```
```suggestion
  let modifiedChartName = chartName.substring(repoName.length). replace (regEx('^/'), '');
```
why not even simpler?

viceice reviewed (commented):
sorry for late comments
